src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -629,73 +632,74 @@ private void SetDefaultProjectName()
 
         private async Task EnsureNuGetAndVsProjectAdapterCacheAsync()
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
-
-            if (!_cacheInitialized && IsSolutionOpen)
+            await _initLock.ExecuteNuGetOperationAsync(async () =>
```

Comment:
in some case, we need this lock, I think this lock should show up in your next PMC/UI PR. because Open solution/Project Add/Project Rename are already protected by SolutionOpenRaised.

this lock is used for UI/PMC race condition.